TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,260 @@
+        &openParams,
+        &symLink,
+        STANDARD_RIGHTS_ALL);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+    }
+
+    return status;
+}
+
+void SetBlackCompletionRoutine(
+    _In_ WDFREQUEST Request,
+    _In_ WDFIOTARGET Target,
+    _In_ PWDF_REQUEST_COMPLETION_PARAMS Params,
+    _In_ WDFCONTEXT Context)
+{
+    UNREFERENCED_PARAMETER(Target);
+    UNREFERENCED_PARAMETER(Params);
+    UNREFERENCED_PARAMETER(Context);
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    status = WdfRequestGetStatus(Request);
+    KdPrint(("TailLight: %s WdfRequestSend status: 0x%x\n", __func__, status));
+
+    // One-shot and top of stack, so delete and pray.
+    WdfObjectDelete(Request);
+}
+
+VOID SetBlackWorkItem(
+    WDFWORKITEM workItem)
+    /*++
+
+    Routine Description:
+
+        Creates a WDF workitem to do the SetBlack() call after the driver
+        stack has initialized.
+
+    Arguments:
+
+        workItem - Handle to a pre-allocated WDF work item.
+    --*/
+{
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+    WDFDEVICE device = static_cast<WDFDEVICE>(WdfWorkItemGetParentObject(workItem));
+    auto workItemContext = WdfObjectGet_SET_BLACK_WORKITEM_CONTEXT(workItem);
+
+    status = SetBlackAsync(device, workItemContext->symLink);
+
+    if (NT_SUCCESS(status)) {
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+        InterlockedIncrement((PLONG)(&pDeviceContext->fSetBlackSuccess));
+    }
+
+    NukeWdfHandle<WDFWORKITEM>(workItem);
+}
+
+NTSTATUS SetBlackAsync(WDFDEVICE device, 
+    CONST UNICODE_STRING& symLink) {
+
+    PAGED_CODE();
+
+    NTSTATUS        status = STATUS_UNSUCCESSFUL;
+    WDFIOTARGET     hidTarget = NULL;
+
+    DEVICE_CONTEXT* pDeviceContext = NULL;
+
+    pDeviceContext =
+        WdfObjectGet_DEVICE_CONTEXT(device);
+
+    if (pDeviceContext == NULL) {
+        return STATUS_DEVICE_NOT_READY;
+    }
+    {
+        WDF_OBJECT_ATTRIBUTES attributes; 
+        WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+
+        // Ensure freed if fails.
+        status = WdfIoTargetCreate(
+            device,
+            &attributes,
+            &hidTarget);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrintEx((DPFLTR_IHVDRIVER_ID,
+                DPFLTR_ERROR_LEVEL,
+                "TailLight: 0x%x while creating I/O target from worker thread\n",
+                status));
+            return status;
+        }
+
+        status = TryToOpenIoTarget(hidTarget, symLink);
+    }
+
+    if (NT_SUCCESS(status)) {
+        
+        UNICODE_STRING theirHwId = {};
+
+        status = AllocateHwIdString(hidTarget, theirHwId);
```

Comment:
`theirHwId` allocations will here pile up causing leakage if `SetBlackAsync` is called multiple times. Could this become an issue?

Reply:
No. Parented to to the IO target.